helusers/oidc.py

```diff
@@ -2,3 +2,131 @@
+        self._issuer = issuer
+
+    def keys(self):
+        config_url = self._issuer + "/.well-known/openid-configuration"
+        config = requests.get(config_url).json()
+
+        keys_url = config["jwks_uri"]
+        return requests.get(keys_url).json()
+
+
+def _build_defaults():
+    class _Defaults:
+        @cached_property
+        def audience(self):
+            if not api_token_auth_settings.AUDIENCE:
+                raise ImproperlyConfigured(
+                    "You must set OIDC_API_TOKEN_AUTH['AUDIENCE'] setting to the accepted JWT audience."
+                )
+            return api_token_auth_settings.AUDIENCE
+
+        @cached_property
+        def issuers(self):
+            issuers = api_token_auth_settings.ISSUER
+            if not issuers:
+                raise ImproperlyConfigured(
+                    "You must set OIDC_API_TOKEN_AUTH['ISSUER'] setting to one or to a list of accepted JWT issuers."
+                )
+            if isinstance(issuers, str):
+                issuers = [issuers]
+            return issuers
+
+        @cached_property
+        def configs(self):
+            configs = dict()
+            for issuer in self.issuers:
+                configs[issuer] = OIDCConfig(issuer)
+            return configs
+
+        @cached_property
+        def key_provider(self):
+            confs = self.configs
+
+            def _key_provider(issuer):
+                return confs[issuer].keys()
+
+            return _key_provider
+
+    return _Defaults()
+
+
+_defaults = _build_defaults()
+
+
+class AuthenticationError(Exception):
+    pass
+
+
+class RequestJWTAuthentication:
+    def __init__(self, key_provider=None):
+        """
+        key_provider: a callable that provides public keys for an issuer. If omitted,
+                      the default is used, which fetches keys from the issuer server
+                      using the OIDC standard configuration URL.
+        """
+        self._key_provider = key_provider or _defaults.key_provider
+
+    def authenticate(self, request):
+        """Looks for a JWT from the request's "Authorization" header. If the header
+        is not found, or it doesn't contain a JWT, returns None.
+
+        If the header is found and contains a JWT then the JWT gets verified.
+        If verification passes, takes a user's id from the JWT's "sub" claim.
+        Creates a User if it doesn't already exist. On success puts the User into
+        request.user and a UserAuthorization object into request.auth and also returns
+        these same objects as a (User, UserAuthorization) tuple.
+        Raises an AuthenticationError on authentication failure."""
+        try:
+            auth_header = request.headers["Authorization"]
+            auth_scheme, jwt_value = auth_header.split()
+            if auth_scheme != "Bearer":
+                return None
+            jwt = JWT(jwt_value)
+        except Exception:
+            return None
```

Comment:
If this is ultimately to be used in the same way as the Django Authentication Middleware, there is a convention to check for authentication like this: https://docs.djangoproject.com/en/3.1/ref/request-response/#django.http.HttpRequest.user

Since this class sets the user in the happy case, maybe it should set `user = AnonymousUser()` in the unhappy one.

I realize this isn't a middleware yet, but maybe worth thinking about?

Reply:
See here: https://github.com/django/django/blob/5ce31d6a7142ca8c76d6b52fa42b3406b9a8ff48/django/contrib/auth/__init__.py#L201

Reply:
Also the @login_required decorator works this way.

Reply:
I don't think this new class itself should _become_ Django Authentication Middleware. But such middleware could well be implemented _using_ this class.

But your comment made me think that perhaps this class shouldn't set anything into the request, but just return the `(user, auth)` tuple. Then the caller does whatever it pleases with those, in the case of Django Authentication Middleware it would return the `User` (because that's what authentication middleware is supposed to do). By not touching the request this service would be "least opinionated" and could potentially be used in many different places.

In any case, if `django.contrib.auth.middleware.AuthenticationMiddleware` is in use, it [sets the request.user](https://github.com/django/django/blob/5ce31d6a7142ca8c76d6b52fa42b3406b9a8ff48/django/contrib/auth/middleware.py#L23) in a way that it defaults to `AnonymousUser` so probably no code needs to ever set `request.user` explicitly to AnonymousUser`.

Reply:
Yeah, good idea. That was kind of my point, since this code already did "one half" of what the Auth middleware does, should it do the other half to for consistency. (For example in cases where AuthenticationMiddleware is not installed.  But you are correct, and returning the tuple would probably be good.

Reply:
Added a commit fixing this. If it looks good, eventually I'll change the commit history so that the user,auth pair is never inserted into the request in the first place. Sound good?

Reply:
Yes,thank you!